coderd/workspaceapps.go

```diff
@@ -119,3 +226,113 @@ func (api *API) workspaceAppsProxyPath(rw http.ResponseWriter, r *http.Request)
 
 	proxy.ServeHTTP(rw, r)
 }
+
+var (
+	// Remove the "starts with" and "ends with" regex components.
+	nameRegex = strings.Trim(httpapi.UsernameValidRegex.String(), "^$")
+	appURL    = regexp.MustCompile(fmt.Sprintf(
+		// {PORT/APP_NAME}--{AGENT_NAME}--{WORKSPACE_NAME}--{USERNAME}
+		`^(?P<AppName>%[1]s)--(?P<AgentName>%[1]s)--(?P<WorkspaceName>%[1]s)--(?P<Username>%[1]s)$`,
+		nameRegex))
+)
+
+// ApplicationURL is a parsed application URL hostname.
+type ApplicationURL struct {
+	// Only one of AppName or Port will be set.
+	AppName       string
+	Port          uint16
+	AgentName     string
+	WorkspaceName string
+	Username      string
+	// BaseHostname is the rest of the hostname minus the application URL part
+	// and the first dot.
+	BaseHostname string
+}
+
+// String returns the application URL hostname without scheme.
+func (a ApplicationURL) String() string {
+	appNameOrPort := a.AppName
+	if a.Port != 0 {
+		appNameOrPort = strconv.Itoa(int(a.Port))
+	}
+
+	return fmt.Sprintf("%s--%s--%s--%s.%s", appNameOrPort, a.AgentName, a.WorkspaceName, a.Username, a.BaseHostname)
+}
+
+// ParseSubdomainAppURL parses an application from the subdomain of r's Host
+// header. If the subdomain is not a valid application URL hostname, returns a
+// non-nil error.
+//
+// Subdomains should be in the form:
+//
+//	{PORT/APP_NAME}--{AGENT_NAME}--{WORKSPACE_NAME}--{USERNAME}
+//	(eg. http://8080--main--dev--dean.hi.c8s.io)
+func ParseSubdomainAppURL(hostname string) (ApplicationURL, error) {
+	subdomain, rest, err := SplitSubdomain(hostname)
+	if err != nil {
+		return ApplicationURL{}, xerrors.Errorf("split host domain %q: %w", hostname, err)
+	}
+
+	matches := appURL.FindAllStringSubmatch(subdomain, -1)
+	if len(matches) == 0 {
+		return ApplicationURL{}, xerrors.Errorf("invalid application url format: %q", subdomain)
+	}
+	matchGroup := matches[0]
+
+	appName, port := AppNameOrPort(matchGroup[appURL.SubexpIndex("AppName")])
+	return ApplicationURL{
+		AppName:       appName,
+		Port:          port,
+		AgentName:     matchGroup[appURL.SubexpIndex("AgentName")],
+		WorkspaceName: matchGroup[appURL.SubexpIndex("WorkspaceName")],
+		Username:      matchGroup[appURL.SubexpIndex("Username")],
+		BaseHostname:  rest,
+	}, nil
+}
```

Comment:
I _think_ `httpapi` might be a better spot for this too, but it's a bit weird, so chef's choice on this one. Regardless, I think it'd be better to remove this from being exposed and use an `_internal.go` test for it.

Reply:
how do you do the internal test thing? I've never seen that before

Reply:
Moved to httpapi